pkg/controller/infinispan/infinispan_controller.go

```diff
@@ -1026,6 +1033,64 @@ func PodEnv(i *infinispanv1.Infinispan, systemEnv *[]corev1.EnvVar) []corev1.Env
+	c := &spec.InitContainers
+	volumes := &spec.Volumes
+	volumeMounts := &spec.Containers[0].VolumeMounts
+	if ispn.HasExternalArtifacts() {
+		extractCommands := externalArtifactsExtractCommand(ispn, ExternalArtifactsMountPath)
+		if position := findInitContainer(*c, ExternalArtifactsDownloadInitContainer); position >= 0 {
+			if spec.InitContainers[position].Args[0] != extractCommands {
+				spec.InitContainers[position].Args = []string{extractCommands}
+				updated = true
+			}
+		} else {
+			*c = append(*c, corev1.Container{
+				Image:   ispn.ImageName(),
+				Name:    ExternalArtifactsDownloadInitContainer,
+				Command: []string{"sh", "-c"},
+				Args:    []string{extractCommands},
+				VolumeMounts: []corev1.VolumeMount{{
+					Name:      ExternalArtifactsVolumeName,
+					MountPath: ExternalArtifactsMountPath,
+				}},
+			})
+			*volumeMounts = append(*volumeMounts, corev1.VolumeMount{Name: ExternalArtifactsVolumeName, MountPath: ExternalArtifactsMountPath, ReadOnly: true})
+			*volumes = append(*volumes, corev1.Volume{Name: ExternalArtifactsVolumeName, VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}})
+			updated = true
+		}
+	} else {
+		if position := findInitContainer(*c, ExternalArtifactsDownloadInitContainer); position >= 0 {
+			volumePosition := findVolume(*volumes, ExternalArtifactsVolumeName)
+			volumeMountPosition := findVolumeMount(*volumeMounts, ExternalArtifactsVolumeName)
+			*c = append(spec.InitContainers[:position], spec.InitContainers[position+1:]...)
+			*volumes = append(spec.Volumes[:volumePosition], spec.Volumes[volumePosition+1:]...)
+			*volumeMounts = append(spec.Containers[0].VolumeMounts[:volumeMountPosition], spec.Containers[0].VolumeMounts[volumeMountPosition+1:]...)
+			updated = true
+		}
+	}
+	return
+}
+
+func externalArtifactsExtractCommand(ispn *infinispanv1.Infinispan, mountPath string) string {
+	extractCommands := fmt.Sprintf("cd %s;", mountPath)
+	for i, artifact := range ispn.Spec.Dependencies.Artifacts {
+		fileName := strings.ToLower(path.Base(artifact.Url))
+		if artifact.Type == infinispanv1.ExternalArtifactTypeZip || (artifact.Type == "" && strings.HasSuffix(fileName, ExternalArtifactsZipExtension)) {
+			downloadFileName := fmt.Sprintf("file%d%s", i, ExternalArtifactsZipExtension)
+			extractCommands = extractCommands + fmt.Sprintf("curl --insecure -L %s -o %s && unzip -oq %s && rm %s;", artifact.Url, downloadFileName, downloadFileName, downloadFileName)
+			continue
+		}
+		if artifact.Type == infinispanv1.ExternalArtifactTypeTarGz || (artifact.Type == "" && strings.HasSuffix(fileName, ExternalArtifactsTarGzExtension)) {
+			downloadFileName := fmt.Sprintf("file%d%s", i, ExternalArtifactsTarGzExtension)
+			extractCommands = extractCommands + fmt.Sprintf("curl --insecure -L %s -o %s && tar xf %s && rm %s;", artifact.Url, downloadFileName, downloadFileName, downloadFileName)
```

Comment:
We can just make this `extractCommands = fmt.Sprintf("%s curl ...` and pass `extractCommands` to `Sprintf`. There's also no need to specify downloadFileName three times, instead you can provide it once and use positional args in the string.

As the `.zip` & `.tar.gz` logic is the same except the extract command, I think we create a function to perform the common operations that simply takes the `artifact.Type`, extension and command as args.

Reply:
Oh, yes, very good point. Thanks